src/webgpu/web_platform/reftests/ref/canvas_use_as_image_source_utils.html.ts

```diff
@@ -0,0 +1,138 @@
+import { assert } from '../../../../common/util/util.js';
+
+export function useAsImageSource2dDrawImage(src: HTMLCanvasElement, dst: HTMLCanvasElement) {
+  const context2d = dst.getContext('2d') as CanvasRenderingContext2D;
+  context2d.drawImage(src, 0, 0);
+}
+
+export function useAsImageSourceToDataURL(src: HTMLCanvasElement, dst: HTMLCanvasElement) {
+  const context2d = dst.getContext('2d') as CanvasRenderingContext2D;
+  const imgFromDataUrl = new Image();
+  imgFromDataUrl.src = src.toDataURL();
+  imgFromDataUrl.onload = () => {
+    context2d.drawImage(imgFromDataUrl, 0, 0);
+  };
+}
+
+export function useAsImageSourceToBlob(src: HTMLCanvasElement, dst: HTMLCanvasElement) {
+  const context2d = dst.getContext('2d') as CanvasRenderingContext2D;
+  const imgFromBlob = new Image(src.width, src.height);
+  src.toBlob(blob => {
+    assert(blob !== null);
+    const url = URL.createObjectURL(blob);
+    imgFromBlob.src = url;
+  });
+  imgFromBlob.onload = () => {
+    context2d.drawImage(imgFromBlob, 0, 0);
+  };
+}
+
+export function useAsImageSourceCreateImageBitmap(src: HTMLCanvasElement, dst: HTMLCanvasElement) {
+  const context2d = dst.getContext('2d') as CanvasRenderingContext2D;
+  createImageBitmap(src).then(image => {
+    context2d.drawImage(image, 0, 0);
+  }, null);
+}
+
+const kVertexShaderSourceWebGL = `
+attribute vec4 a_Position;
+varying vec2 v_Texcoord;
+void main()
+{
+  v_Texcoord = (a_Position.xy + vec2(1.0, 1.0)) * 0.5;
+  v_Texcoord.y = 1.0 - v_Texcoord.y;
+  gl_Position = a_Position;
+}
+`;
+
+const kFragmentShaderSourceWebGL = `
+precision mediump float;
+varying vec2 v_Texcoord;
+uniform sampler2D u_Texture2D;
+void main() {
+  gl_FragColor = texture2D(u_Texture2D, v_Texcoord);
+}
+`;
+
+const kVertices = new Float32Array([1, 1, -1, 1, -1, -1, 1, 1, -1, -1, 1, -1]);
+
+function WebGLDrawTexQuad(gl: WebGLRenderingContext) {
```

Comment:
note: If we switch to non-reftests - might be able to use readPixels instead of drawing a quad.